packages/caravan-fees/src/btcTransactionTemplate.ts

```diff
@@ -0,0 +1,609 @@
+   * @returns True if a change output is needed, false otherwise
+   */
+  get needsChangeOutput(): boolean {
+    const targetFeesWithDustBuffer = this.targetFees().plus(
+      this._dustThreshold,
+    );
+
+    return (
+      !this.malleableOutputs.length &&
+      this.calculateCurrentFee().gt(targetFeesWithDustBuffer)
+    );
+  }
+
+  /**
+   * Calculates the total input amount.
+   * @returns {Satoshis} The total input amount in satoshis (as a string)
+   */
+  getTotalInputAmount(): Satoshis {
+    return this.calculateTotalInputAmount().toString();
+  }
+
+  /**
+   * Calculates the total change amount. (Total Inputs Amount - Total Non-change (non-malleable) Outputs Amount )
+   * @returns {Satoshis} The total change amount in satoshis (as a string)
+   */
+  get changeAmount(): Satoshis {
+    return this.calculateChangeAmount().toString();
+  }
+
+  /**
+   * Calculates the total output amount.
+   * @returns {Satoshis} The total output amount in satoshis (as a string)
+   */
+  getTotalOutputAmount(): Satoshis {
+    return this.calculateTotalOutputAmount().toString();
+  }
+
+  /**
+   * Estimates the virtual size of the transaction.
+   * @returns The estimated virtual size in vbytes
+   */
+  get estimatedVsize(): number {
+    return estimateTransactionVsize({
+      addressType: this._scriptType,
+      numInputs: this._inputs.length,
+      numOutputs: this._outputs.length,
+      m: this._requiredSigners,
+      n: this._totalSigners,
+    });
+  }
+
+  /**
+   * Calculates the estimated fee rate of the transaction.
+   * @returns {string} The estimated fee rate in satoshis per vbyte
+   */
+  get estimatedFeeRate(): string {
+    return this.calculateCurrentFee()
+      .dividedBy(this.calculateEstimatedVsize())
+      .toString();
+  }
+
+  /**
+   * Adds an input to the transaction.
+   * @param input - The input to add
+   */
+  addInput(input: BtcTxInputTemplate): void {
+    this._inputs.push(input);
+  }
+
+  /**
+   * Adds an output to the transaction.
+   * @param output - The output to add
+   */
+  addOutput(output: BtcTxOutputTemplate): void {
+    this._outputs.push(output);
+  }
+
+  /**
+   * Removes an output from the transaction.
+   * @param index - The index of the output to remove
+   */
+  removeOutput(index: number): void {
+    this._outputs.splice(index, 1);
+  }
+
+  /**
+   * Adjusts the change output of the transaction.
+   * This method calculates a new change amount based on the current inputs,
+   * non-change outputs, and the target fee. It then updates the change output
+   * or removes it if the new amount is below the dust threshold.
+   *
+   * The method also handles cases where the change output already has an amount:
+   * 1. It calculates the difference between the new and current change amount.
+   * 2. If removing change, it ensures the fee doesn't unexpectedly increase.
+   */
+  adjustChangeOutput(): void {
+    if (this.malleableOutputs.length === 0) return;
```

Comment:
> Should this throw an error? If this method is called presumably there's an expectation that change needs to be adjusted. if it just returns (which is the same that happens when an adjustment happens successfully, you could either end up in an infinite loop if you're expecting results to change after calling this or potentially overpaying fees if you just assume that a return without erroring means it worked when in truth there's no adjusted change to account for excess fees.


Thanks for pointing that out, this can indeed lead to silent failures, without one realizing where the error originated from

so what I was thinking of was that what if instead of throwing an error (which I think is too aggressive here)  we returned a boolean (indicating that any adjustments were done or not) , why I prefer this over throwing an error is because the `adjustChangeOutput` is supposed to run in the RBF and CPFP creation process and the developer using it might not be sure that if he has added any change outputs (or the algorithm we have set have added any), so it could lead to forcibly interrupting the flow of the program. 
So if we returned false, the user code can then add a malleable output maybe and then re-run the `adjustChangeOutput` fn ....  

so what do you think ??

Reply:
What if instead of a bool it returned the new amount or null/void?

Reply:
changes [done](https://github.com/caravan-bitcoin/caravan/pull/114/commits/9b4a1746c88d3ee0e7b61b9d29af83fb93445934#diff-c113a33821d4e86e4ad4f20743227a51337831792cc5c5926aa2a5971c11d7b1R239)